src/blockdev.rs

```diff
@@ -105,37 +141,58 @@ impl Disk {
+        let mut result: Vec<Partition> = Vec::new();
+        let isblktypeavailable: bool = self.validate_blktype(&devinfo.blktype, with_holders);
+        if isblktypeavailable {
+            let (mountpoint, swap) = match &devinfo.mountpoint {
+                Some(mp) if mp == "[SWAP]" => (None, true),
+                Some(mp) => (Some(mp.to_string()), false),
+                None => (None, false),
+            };
+            result.push(Partition {
+                path: devinfo.name.to_owned(),
+                label: devinfo.label.clone(),
+                fstype: devinfo.fstype.clone(),
+                parent: self.path.to_owned(),
+                mountpoint,
+                swap,
+            });
+        }
+        Ok(result)
+    }
+
     fn get_partitions(&self, with_holders: bool) -> Result<Vec<Partition>> {
         // walk each device in the output
         let mut result: Vec<Partition> = Vec::new();
-        for devinfo in lsblk(Path::new(&self.path), true)? {
-            if let Some(name) = devinfo.get("NAME") {
-                match devinfo.get("TYPE") {
-                    // If unknown type, skip.
-                    None => continue,
-                    // If whole-disk device, skip.
-                    Some(t) if t == &"disk".to_string() => continue,
-                    // If partition, allow.
-                    Some(t) if t == &"part".to_string() => (),
-                    // If with_holders is true, allow anything else.
-                    Some(_) if with_holders => (),
-                    // Ignore LVM or RAID devices which are using one of the
-                    // partitions but aren't a partition themselves.
-                    _ => continue,
-                };
-                let (mountpoint, swap) = match devinfo.get("MOUNTPOINT") {
-                    Some(mp) if mp == "[SWAP]" => (None, true),
-                    Some(mp) => (Some(mp.to_string()), false),
-                    None => (None, false),
-                };
-                result.push(Partition {
-                    path: name.to_owned(),
-                    label: devinfo.get("LABEL").map(<_>::to_string),
-                    fstype: devinfo.get("FSTYPE").map(<_>::to_string),
-                    parent: self.path.to_owned(),
-                    mountpoint,
-                    swap,
-                });
+        let deviceinfos: Vec<Device> = Device::lsblk(Path::new(&self.path), true)?;
+        for deviceinfo in deviceinfos {
+            let mut partition = self.compute_partition(&deviceinfo, with_holders)?;
+            result.append(&mut partition);
+            if let Some(children) = deviceinfo.children.as_ref() {
+                if !children.is_empty() {
+                    for child in children {
+                        let mut childpartitions = self.compute_partition(child, with_holders)?;
+                        result.append(&mut childpartitions)
+                    }
+                }
```

Comment:
Hmm, right OK you had to do some refactoring here to deal with children not being in the main iteration anymore.

Though I don't think this flow is functionally equivalent, because we're only going one level down. I think we want a recursive model here instead? So e.g. you have `get_partitions` call to `get_partitions_recurse` with the top-level `Device`, and then that function calls itself for each child in the device, aggregating partitions in the vector as it goes. Then we return that vector. Does that make sense?